content/blog/2023/12/camel-data-types/index.md

```diff
@@ -0,0 +1,490 @@
+   dataTypes:
+     out:
+       format: aws2-s3:application-cloudevents # (1)
+ sink:
+   ref:
+     kind: Kamelet
+     apiVersion: camel.apache.org/v1
+     name: http-sink
+   dataTypes:
+     in:
+       format: http:application-cloudevents  # (2)
+   properties:
+     url: "{{http.sink.url}}"
+```
+
+1. Use aws2-s3 specific CloudEvents output which sets specific headers
+2. Use Http binding of CloudEvents to produce a proper Http request following the CloudEvents specification
+
+The Pipe is able to choose input and output data types when referencing Kamelets as sources and sinks. 
+This way you can adapt the Kamelets behavior and the combination of Kamelets becomes much more robust as the Kamelets are able to work hand in hand with increased interoperability.
+
+Another example shows that Kamelet data types also help to transform the message as part of a Pipe. 
+There are various Kamelets available to transform JsonNode based workloads such as extract-field-action, insert-field-action, hoist-field-action. 
+All these action Kamelets operate on a generic JsonNode model object. 
+The data types help to transform the Kamelet source output into such a generic JsonNode by using the _"application-x-struct"_ data type.
+
+```yaml
+apiVersion: camel.apache.org/v1
+kind: Pipe
+metadata:
+ name: slack-to-kafka
+spec:
+ source:
+   ref:
+     kind: Kamelet
+     apiVersion: camel.apache.org/v1
+     name: slack-source
+   properties:
+     channel: "{{slack.channel}}"
+     token: "{{slack.bot.token}}"
+   dataTypes:
+     out:
+       format: application-x-struct # (1)
+ steps:
+   - ref:
+       kind: Kamelet
+       apiVersion: camel.apache.org/v1
+       name: extract-field-action
+     properties:
+       field: text # (2)
+ sink:
+   ref:
+     kind: KafkaTopic
+     apiVersion: kafka.strimzi.io/v1beta2
+     name: "{{kafka.topic}}"
+   dataTypes:
+     in:
+       format: plain-text # (3)
+   properties:
+     brokers: "{{kafka.bootstrapServers}}"
+```
+
+1. Use "application-x-struct" data type that transforms the Slack Json into a generic JsonNode using the Jackson library
+2. Extract the "text" field (user input on the Slack channel) from the JsonNode and set as a new message body
+3. Transform the message to "plain-text" before sending the data to Kafka
+
+This example Pipe shows that data types tremendously improve the interaction between Kamelets because it is very easy to transform the message content into different data formats as part of the Pipe processing by just referencing some data types.
+
+The input/output data types on Kamelets perfectly complement the declarative nature of Kamelet and Pipes.
+
+# What’s next!?
+
+The data types feature is available since Camel 4 and the next step is to provide more data type transformer implementations for different types of Camel components. 
+So please do not hesitate to provide feedback and get involved by adding more data types that everybody can use.
```

Comment:
Is there already a list of Data type transformers ideas somewhere? In Jira?

providing one and linking it from here could really help to have people starting to provide some.

Reply:
good idea! unfortunately there is no such list and it is not easy for me to create one in the short term